### PR TITLE
Add Offset.reset() and remove unused Offset implementations

### PR DESCRIPTION
**Pull request by leventov**
- Add `Offset.reset()`, which will allow to create ColumnValueSelectors/DimensionSelectors against a static `Offset` object (currently "reset" is done via `Offset` cloning and replacing the `Offset` object), that would allow to decrease the required level of indirection: instead of referencing the object, which contains a reference to `Offset` (`OffsetHolder`, in fact), selector will be able to reference the offset directly (which is already a "holder" of primitive `int` offset value). It will also make "queryable" and "incremental" index implementations more aligned with each other (in the incremental index, there is `IncrementalIndex.EntryHolder` which references changing `TimeAndDims`)

 - Remove unused `ArrayBasedOffset`, `IntersectingOffset` and `UnioningOffset`.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/BitmapOffset.java [hunk 1]
```diff
@@ -123,12 +122,13 @@ private static String factorizeFullness(long bitmapCardinality, long numRows)
...
+  private final String fullness;
+  private IntIterator iterator;
+  private final IntIterator iteratorForReset;
+  private final int valueForReset;
+  private int value;
 
-  int val;
-
-  public static IntIterator getReverseBitmapOffsetIterator(ImmutableBitmap bitmapIndex)
+  static IntIterator getReverseBitmapOffsetIterator(ImmutableBitmap bitmapIndex)
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/BitmapOffset.java]: Why no longer public?

File: processing/src/main/java/io/druid/segment/BitmapOffset.java [hunk 2]
```diff
@@ -169,65 +167,65 @@ private IntIterator newIterator(ImmutableBitmap bitmapIndex, boolean descending)
...
-      super(fullness, itr, val);
-    }
-
-    @Override
-    public Offset clone()
-    {
-      return new RoaringBitmapOffset(fullness, itr.hasNext() ? itr.clone() : EmptyIntIterator.instance(), val);
-    }
+    // Calling clone() on empty iterators from RoaringBitmap library sometimes fails with NPE,
+    // see https://github.com/RoaringBitmap/RoaringBitmap/issues/177
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/BitmapOffset.java]: Do we need to file an issue / PR for this?

File: processing/src/main/java/io/druid/segment/data/Offset.java [hunk 3]
```diff
@@ -40,6 +40,11 @@
   @CalledFromHotLoop
   public abstract boolean withinBounds();
 
+  /**
+   * Resets the Offset to the position it was created or cloned with.
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/data/Offset.java]: Reset is not thread safe, suggest adding in explicit callout to thread safety either in the class description or at least in this method.

Comment (leventov) [inline processing/src/main/java/io/druid/segment/data/Offset.java]: Added a note to the class description

File: processing/src/main/java/io/druid/segment/BitmapOffset.java (hunk 1 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/BitmapOffset.java]: It didn't need to be public, there was a warning in IntelliJ

File: processing/src/main/java/io/druid/segment/BitmapOffset.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/BitmapOffset.java]: Added #4709

**Pull request merged**